src/unix/syscall.c

```diff
@@ -2066,8 +2066,10 @@ sysreturn exit_group(int status)
     threads_to_vector(current->p, v);
     spin_unlock(&p->threads_lock);
 
+    shutting_down = true;
+    wakeup_or_interrupt_cpu_all();
     vector_foreach(v, t)
-            exit_thread(t);
+        exit_thread(t);
```

Comment:
If the sending of wakeup IPIs is asynchronous, could there possibly be a race here where a thread's frame is deallocated and yet it gets scheduled on another core before the IPI is caught? It kind of seems unlikely - but also that there's no guarantee that it couldn't happen.

Is there really a need to call exit_thread at all if the system is shutting down and threads can't get scheduled anyway?

Reply:
Setting shutting_down before bumping the other cpus to runloop should prevent them from being scheduled before tearing them down, since the other places to get scheduled should be protected by the kernel lock or a check for shutting_down, but thinking about it some more I'm less certain about what happens when another cpu is in the middle of random interrupt.
I was assuming that the teardown and flushing of queues that happens in exit_thread was necessary, but I suppose if the user program is calling exit_group then it has already made sure that all the work it needs done is complete. Eliminating the calls to exit_thread is a much simpler and more reliable way to handle any possible races.